### PR TITLE
Add APM server and APM test agent to compose file

### PR DESCRIPTION
**Pull request by HarisSistek**
Unsure if this of interest to you, but was doing some prototyping with Elasticsearch APM server using your elk stack and wanted to return the favor if an APM server is wanted in the stack. 

Why:
* Adds Elasticsearch Application Performance Monitoring
* Can view application traces, errors and alerts trough Kibana

How:
* Add apm-server container to the network which will receive traffic on port 8200 and report to elastic search server at port 9200
* Add a test flask app with redis to view apm tracing

### CODE REVIEW

**Comment by deviantony** [comment]
IMO should be added as an extension rather than as a part of the core, see https://github.com/deviantony/docker-elk#how-can-i-enable-the-provided-extensions

**Comment by HarisSistek** [comment]
@deviantony good suggestion, much cleaner configuration this way. Moved it to two extensions, one for the apm server and one for the apm agent.

**Comment by deviantony** [comment]
@HarisSistek why packaging it under two extensions instead of one?

**Comment by HarisSistek** [comment]
Maybe easier to explain with this image:
image

The APM agents and beats are meant to be at hosts and send data to their respective servers. 

The `apm-agent-compose.yml` was just meant as en example flask app on how to connect an apm agent to the apm server and send traces. I could remove the apm-agent from `extensions` dir to perhaps a new dir called `example-connectors` that shows how to connect beats and apm-agents.

What do you suggest?

**Comment by deviantony** [comment]
Ok, thanks for the illustrated answer. Makes sense. 

Similar to the fact that we do not have a beats extension, I think that we should remove the apm-agent extension and instead document the connection from agent to server and how to send traces in the Wiki instead (in a new "Connectors" section for example).

Thoughts @antoineco ?

**Comment by HarisSistek** [comment]
@deviantony @antoineco So to recap, I will:
- remove the apm-agent
- add description on how to connect apm agents to the apm server in the extensions/apm-server/readme

About making pre-buildt image and demo app: Unsure how much time I have to make them. Also they should probably be made under the deviantony dockerhub domain so people know that they are related  to the github repo. E.g. deviantony/elk-demo-apm-agent

**Comment by antoineco** [comment]
Sounds good to me!

Could we apply the agent instructions to a generic component that's already available on the Docker Hub? (bear with me, I've never used APM)

**Comment by HarisSistek** [comment]
Well the apm agent has to wrap the running process and forward data to the apm-server. 

e.g. java:
```
java \
 -javaagent:/.../elastic-apm-agent-<version>.jar\
 -Delastic.apm.service_name=my-application\
 -Delastic.apm.server_url=http://localhost:8200\
 -Delastic.apm.application_packages=org.example\
 -jar my-application.jar
```

while nodejs:
```
// Add this to the VERY top of the first 
// file loaded in your app
var apm = require('elastic-apm-node').start({
  // Override service name from package.json 
  // Allowed characters: a-z, A-Z, 0-9, -, _, 
  // and space
  serviceName: '',
  
  // Use if APM Server requires a token
  secretToken: '',
  
  // Set custom APM Server URL
  // Default: http://localhost:8200
  serverUrl: ''
})
``` 

I haven't looked in dockerhub if there are any ready examples. You can read more about the elk apms here: https://www.elastic.co/solutions/apm and each language has its own apm doc that explains how to use it.

**Comment by HarisSistek** [comment]
@deviantony @antoineco I removed the apm-agent, and added some more information on the apm-server/readme on connecting, finding apm agent docs etc. Do tell if you want anything changed

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+.idea/*
```
Comment (antoineco) [inline .gitignore]: Not sure if we really need that, this project is essentially Dockerfiles and a YAML file.

File: extensions/apm-server/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,3 @@
+ARG ELK_VERSION
+
+FROM docker.elastic.co/apm/apm-server:${ELK_VERSION}
```
Comment (antoineco) [inline extensions/apm-server/Dockerfile]: Please make sure all committed files end with a newline character ([POSIX compliance](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206))

> Preferences > Editor > General > Ensure line feed at file end on save

File: .gitignore (hunk 1 above)
Comment (deviantony) [reply .gitignore]: Agree, no need to add the .gitignore file.

File: extensions/apm-server/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    'DEBUG': True,
+}
+```
+
+More configuration setting can be found under the **Configuration**
+section for each language. Link: https://www.elastic.co/guide/en/apm/agent/index.html 
+
+## Checking Connectivity and Importing default APM Dashboards
+
+From Kibanas main window press:
```
Comment (deviantony) [inline extensions/apm-server/README.md]:
```suggestion
From Kibana main window press:
```

**Comment by HarisSistek** [comment]
@deviantony @antoineco thx for the feedback, here you go.

**Comment by deviantony** [comment]
Thanks @HarisSistek :+1:

**Comment by HarisSistek** [comment]
@deviantony @antoineco No problem, just repaying the favor. Your stack helped a lot while prototyping

**Pull request merged**